FEATURE: Support SQL Server (and more?)

kdambekalns opened the pull request:
This switches the hard-coded SQL against Doctrine DBAL query builder use.

This makes the queue work on SQL Server and potentially more databases supported by Doctrine DBAL.

kdambekalns committed:
TASK: Code cleanup

- add return types
- add types to method arguments
- drop unneeded doc comments

kdambekalns committed:
FEATURE: Support SQL Server (and more?)

This switches the hard-coded SQL against Doctrine DBAL query builder
use.

This makes the queue work on SQL Server and potentially more databases
supported by Doctrine DBAL.

kdambekalns commented on README.md:
```diff
@@ -63,28 +63,20 @@ The `DoctrineQueue` supports following options:
 |--------|---------|--------:|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | delay  | integer |       0 | Number of seconds before a message is marked "ready" after submission. This can be useful to prevent premature execution of jobs (i.e. before entities are persisted) |
 
-### Release options
-
-Additional options to be specified via `releaseOptions`: 
-
-| Option | Type    | Default | Description                                                                      |
-|--------|---------|--------:|----------------------------------------------------------------------------------|
-| delay  | integer |       0 | Number of seconds before a message is marked "ready" after it has been released. |
-
```
This is part of the `JobQueue.Common` package and documented there.

kdambekalns replied:
Ok, I just realized the documentation there says

> Implementation specific (see corresponding package)

I'll bring it back here…

kdambekalns commented on Tests/Functional/Queue/DoctrineQueueTest.php:
```diff
@@ -26,6 +26,6 @@ class DoctrineQueueTest extends AbstractQueueTest
     protected function getQueue()
     {
-        return new DoctrineQueue('Test-queue', $this->queueSettings);
+        return new DoctrineQueue('test-queue', $this->queueSettings);
```
That fixed the tests for mw, MariaDB would complain about a non-existing table, because of case differences between `flowpack_jobqueue_messages_Test-queue` and `flowpack_jobqueue_messages_test-queue`.

Not really a "proper" fix for that, but I was focused on something entirely different… 😎

bwaidelich replied:
I'm still puzzled, that this failed for you since MariaDB should be case insensitive regarding table and column names – at least I _thought_ it was

kdambekalns replied:
I looked at the docs and it wasn't as clear as I hoped it to be. 🤷

kitsunet replied:
quoting
https://dev.mysql.com/doc/refman/8.0/en/identifier-case-sensitivity.html

> Consequently, the case sensitivity of the underlying operating system plays a part in the case sensitivity of database, table, and trigger names. This means such names are not case-sensitive in Windows, but are case-sensitive in most varieties of Unix....

kdambekalns replied:
Ah, I was probably mislead by the fact that my MariaDB runs "on Linux" (in Docker), but uses the host filesystem for storage (on a Mac, case-insensitive). 🙄

kitsunet replied:
Maybe we should just strtolower it just to be on the safe side?

kdambekalns replied:
Having it lowercase in the test code is fine, IMHO. Most people probably use lowercase dientifers for thir queues anyway…

kdambekalns reviewed (commented):
I tested this locally against MariaDB 10.5, SQLite and SQL Server 2022.

bwaidelich commented on Classes/Queue/DoctrineQueue.php:
```diff
@@ -191,15 +200,38 @@ protected function reserveMessage(?int $timeout = null): ?Message
         }
         $this->reconnectDatabaseConnection();
 
+        $qb = $this->connection->createQueryBuilder();
```
We could reduce some duplication with a
```php
private function createQueryBuilder(): QueryBuilder
{
    return $this->connection->createQueryBuilder()
        ->from($this->tableName)
        ->orderBy('id');
}
```

kdambekalns replied:
While `orderBy('id')` appears only twice, a "method for `from()`" can be used six times. 👍

bwaidelich commented on Classes/Queue/DoctrineQueue.php:
```diff
@@ -191,15 +200,38 @@ protected function reserveMessage(?int $timeout = null): ?Message
+            ->from($this->connection->quoteIdentifier($this->tableName))
+            ->where(
+                $qb->expr()->and(
+                    $qb->expr()->eq('state', $qb->expr()->literal('ready')),
```
I'm wondering about the many literal comparisons.
Why not
```php
->andWhere('state = :state')->setParameter('state', 'ready');
```
?

kdambekalns replied:
Good point. I'm doing too little "direct DB stuff"… 😬

bwaidelich commented on Classes/Queue/DoctrineQueue.php:
```diff
@@ -191,15 +200,38 @@ protected function reserveMessage(?int $timeout = null): ?Message
             }
             if ($row !== false) {
-                $numberOfUpdatedRows = (int)$this->connection->executeStatement("UPDATE {$this->connection->quoteIdentifier($this->tableName)} SET state = 'reserved' WHERE id = :id AND state = 'ready' AND {$this->getScheduledQueryConstraint()}", ['id' => (int)$row['id']]);
+                $innerQueryBuilder = $this->connection->createQueryBuilder();
```
why is that one "inner"?

kdambekalns replied:
It's in a loop and I need the "original" `$qb` to stay intact.

kdambekalns replied:
I'lll move out out of the loop…

bwaidelich commented on Classes/Queue/DoctrineQueue.php:
```diff
@@ -301,24 +376,16 @@ protected function resolveScheduledQueryPart(array $options): string
-            default:
-                return '(scheduled IS NULL OR scheduled <= NOW())';
-        }
+        return $qb->expr()->or(
```
Maybe you didn't want to change too much, but this could as well mutate the query builder (like we do in many places for the Neos 9 projections, e.g. in the [ContentSubgraph](https://github.com/neos/neos-development-collection/blob/9.0/Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php#L508))

```
protected function addScheduledQueryConstraints(QueryBuilder $qb): void
{
    $qb->andWhere(
        $qb->expre()->or(
            $qb->expr()->isNull('scheduled'),
            $qb->expr()->lte('scheduled', $this->connection->getDatabasePlatform()->getCurrentTimestampSQL())
        )
    );
}
```

but I'm not sure if that's any better really :)

kdambekalns replied:
I'll leave that one as it is… 🤷

bwaidelich commented on Classes/Queue/DoctrineQueue.php:
```diff
@@ -301,24 +376,16 @@ protected function resolveScheduledQueryPart(array $options): string
-        }
+        return $qb->expr()->or(
+            $qb->expr()->isNull('scheduled'),
+            $qb->expr()->lte('scheduled', $this->connection->getDatabasePlatform()->getCurrentTimestampSQL())
```
Kewl, I didn't know that

kdambekalns replied:
Me neither, until now… 🙃

bwaidelich reviewed (commented):
Nice one, great addition!
No time to test atm, but I left some (partly rather unqualified) comments already

kdambekalns commented:
BTW, this is (slightly) related, it adds a test that invokes a formerly untested code path: https://github.com/Flowpack/jobqueue-common/pull/62

kdambekalns committed:
TASK: Tweaks after code review

- simplify query builder use
- add some sprintf
- rename some variables
- amend docblocks

kdambekalns commented:
Thanks @bwaidelich – I improved the code!

bwaidelich reviewed (approved):
Looking great, +1 by reading!

kdambekalns commented:
Have been testing this the last two days, LGTM.

bwaidelich commented on Classes/Queue/DoctrineQueue.php:
```diff
@@ -27,81 +31,56 @@
         }
         if (isset($options['pollInterval'])) {
-            $this->pollInterval = (int)($options['pollInterval'] * 1000000);
+            $this->pollInterval = (int)$options['pollInterval'] * 1000000;
```
This was a regression (see #31)

kdambekalns replied:
Sub-second intervals… I did not test that! 😬

kdambekalns replied:
See https://github.com/Flowpack/jobqueue-doctrine/pull/32
